cmd/eval-dev-quality/cmd/evaluate.go

```diff
@@ -501,6 +499,68 @@ func (command *Evaluate) evaluateDocker(ctx *evaluate.Context) (err error) {
 	}
 	parallel.Wait()
 
+	// Copy data from volume back.
+	{
+		// Run a container mounting the volume.
+		output, err := util.CommandWithResult(context.Background(), command.logger, &util.Command{
+			Command: []string{
+				"docker",
+				"run",
+				"-d",
+				"--name", "volume-fetch",
+				"-v", volumeName + ":/data",
+				"busybox",
+				"true",
+			},
+		})
+		if err != nil {
+			return pkgerrors.WithMessage(pkgerrors.WithStack(err), output)
+		}
+
+		// Cleanup volume mount container.
+		containerCleanup := func() {
+			output, deferErr := util.CommandWithResult(context.Background(), command.logger, &util.Command{
+				Command: []string{
+					"docker",
+					"rm",
+					"volume-fetch",
+				},
+			})
+			if deferErr != nil {
+				err = errors.Join(err, pkgerrors.WithMessage(pkgerrors.WithStack(deferErr), output))
+			}
+		}
+
+		// Copy data from volume to filesystem.
+		output, err = util.CommandWithResult(context.Background(), command.logger, &util.Command{
+			Command: []string{
+				"docker",
+				"cp",
+				"volume-fetch:/data/.",
+				command.ResultPath,
+			},
+		})
+
+		// Cleanup volume.
+		volumeCleanup := func() {
+			output, deferErr := util.CommandWithResult(context.Background(), command.logger, &util.Command{
+				Command: []string{
+					"docker",
+					"volume",
+					"rm",
+					volumeName,
+				},
+			})
+			if deferErr != nil {
+				err = errors.Join(err, pkgerrors.WithMessage(pkgerrors.WithStack(deferErr), output))
+			}
+		}
+
+		// Defer the cleanup in the right order, the container needs to be removed before removing the volume.
+		defer volumeCleanup()
+		defer containerCleanup()
+	}
```

Comment:
would move the cleanup defers up to where the conainter or data volume is created to ensure 100% it is removed when intermediate errors happen